src/coreclr/src/vm/multicorejitplayer.cpp

```diff
@@ -1120,6 +1132,181 @@ HRESULT MulticoreJitProfilePlayer::HandleMethodRecord(unsigned * buffer, int cou
 }
 
+HRESULT MulticoreJitProfilePlayer::HandleNDirectStubRecord(unsigned * buffer, int count)
+{
+    STANDARD_VM_CONTRACT;
+
+    HRESULT hr = E_ABORT;
+
+    MulticoreJitTrace(("NDirectStubRecord(%d) start %d methods, %d mod loaded", m_stats.m_nTotalMethod, count, m_nLoadedModuleCount));
+
+    MulticoreJitManager & manager = GetAppDomain()->GetMulticoreJitManager();
+
+#ifdef MULTICOREJIT_LOGGING
+
+    MulticoreJitCodeStorage & curStorage = manager.GetMulticoreJitCodeStorage();
+
+    int lastCompiled = curStorage.GetStored();
+
+#endif
+
+    int pos = 0;
+
+    while (! ShouldAbort(true) && (pos < count))
+    {
+        unsigned stubInfo = buffer[pos]; // moduleIndex + methodIndex
+
+        unsigned moduleIndex = stubInfo >> 24;
+
+        if (moduleIndex < m_moduleCount)
+        {
+            PlayerModuleInfo & info = m_pModules[moduleIndex];
+
+            m_stats.m_nTotalMethod ++;
+
+            // If module is disabled for Jitting, just skip method without even waiting
+            if (! info.m_enableJit)
+            {
+                m_stats.m_nFilteredMethods ++;
+            }
+            else
+            {
+                //  To reduce contention with foreground thread, walk backward within the group of methods Jittable methods, not broken apart by dependency
+                {
+                    int run = 1; // size of the group
+
+                    while (((pos + run) < count) && MethodJifInfo(buffer[pos + run]))
+                    {
+                        run ++;
+
+                        // If walk-back run is too long, lots of methods in the front will be missed by background thread
+                        if (run > MAX_WALKBACK)
+                        {
+                            break;
+                        }
+                    }
+
+                    if (run > 1)
+                    {
+                        MulticoreJitTrace(("Jit backwards %d methods",  run));
+                    }
+
+                    // Walk backwards within the same group, may be from different modules
+                    for (int p = pos + run - 1; p >= pos; p --)
+                    {
+                        unsigned inst = buffer[p];
+
+                        _ASSERTE(MethodJifInfo(inst));
+
+                        PlayerModuleInfo & mod = m_pModules[inst >> 24];
+
+                        if (mod.IsModuleLoaded() && mod.m_enableJit)
+                        {
+                            unsigned methodIndex = inst & METHODINDEX_MASK;
+                            unsigned token = TokenFromRid(methodIndex, mdtMethodDef);
+                            MethodDesc * pMethod = MemberLoader::GetMethodDescFromMemberDefOrRefOrSpec(mod.m_pModule, token, NULL, FALSE, FALSE);
+                            if (pMethod != NULL && !pMethod->HasILHeader())
+                            {
+                                GetStubForInteropMethod(pMethod);
+                                pMethod->GetOrCreatePrecode();
```

Comment:
Is the `GetOrCreatePrecode()` call necessary here? I do not look right.

Reply:
Not necessary. I want to do as much as possible in multicorejitplayer.
I will fix it. Thank you.